sdk/src/trace/samplers/probability.cc

```diff
@@ -0,0 +1,118 @@
+// Copyright 2020, Open Telemetry Authors
+// Copyright 2017, OpenCensus Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "opentelemetry/sdk/trace/samplers/probability.h"
+
+#include <cmath>
+#include <cstdint>
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace sdk
+{
+namespace trace
+{
+namespace trace_api = opentelemetry::trace;
+
+ProbabilitySampler::ProbabilitySampler(
+  double probability, bool defer_parent, SamplingBehavior sampling_behavior 
+)
+: threshold_(CalculateThreshold(probability)), probability_(probability),
+  defer_parent_(defer_parent), sampling_behavior_(sampling_behavior) {}
+
+SamplingResult ProbabilitySampler::ShouldSample(
+  const SpanContext *parent_context,
+  trace_api::TraceId trace_id,
+  nostd::string_view name,
+  trace_api::SpanKind span_kind,
+  const trace_api::KeyValueIterable &attributes) noexcept
+{
+  // Check if current Span should be sampled with current Sampling Behavior
+  if (sampling_behavior_ == SamplingBehavior::DETACHED_SPANS_ONLY
+      && parent_context != nullptr && !parent_context->is_remote) {
+    return { Decision::NOT_RECORD, nullptr };
+  }
+
+  if (sampling_behavior_ == SamplingBehavior::ROOT_SPANS_ONLY
+      && parent_context != nullptr) {
+    return { Decision::NOT_RECORD, nullptr };
+  }
+
+  // Check if sampling decision should be defered to parent span, if applicable 
+	if (parent_context == nullptr || !defer_parent_)
+  {
+    if (threshold_ == 0) return { Decision::NOT_RECORD, nullptr };
+
+    // Calculcate probability based decision based on trace_id and threshold
+	  if (CalculateThresholdFromBuffer(trace_id) <= threshold_)
+    {
+      return { Decision::RECORD_AND_SAMPLE, nullptr };
+    }
+	}
+  else
+  {
+    // Return sampling decision based on parent span_context config
+    if (parent_context->sampled)
+    {
+      return { Decision::RECORD_AND_SAMPLE, nullptr };
+    }
+	}
+
+  return { Decision::NOT_RECORD, nullptr };
+}
+
+std::string ProbabilitySampler::GetDescription() const noexcept
+{
+	char buffer[30];
+	sprintf(buffer, "ProbabilitySampler{%.6f}", GetProbability());
+  return std::string(buffer);
```

Comment:
I modified function to cache the string, but I think we'll need to create a separate PR to change the data type to a string_view (since I'll end up breaking the other samplers if I change the sampler.h here)